PR: https://github.com/odoo/odoo/pull/281733

From: f40171000f66ee0371cb50d605ff3517d5235556
From: Stéphane Diez
Date: 2026-08-27 00:36:00

Breaking data model changes scores: del:2 + add:0, change matches:
-    based_on = fields.Selection(
-    percent_factor = fields.Integer(default=100, required=True)

Total Changes: 62

[ADD] stock: orderpoint suggest wizard

Add a new wizard to mass rewrite `based_on` and `percent_factor` on
multiple orderpoints. This will also rewrite the daily demand, the min
qty & the max qty of the orderpoint.

task 6304907

Part-of: odoo/odoo#281733
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/stock/models/stock_orderpoint.py
+++ b/addons/stock/models/stock_orderpoint.py
@@ -5,7 +5,7 @@ from collections import defaultdict
 from datetime import datetime, time, UTC
 from zoneinfo import ZoneInfo
 
-from dateutil import relativedelta
+from dateutil.relativedelta import relativedelta
 from psycopg2 import OperationalError
 
 from odoo import SUPERUSER_ID, _, api, fields, models
@@ -67,7 +67,7 @@ class StockWarehouseOrderpoint(models.Model):
              "reached by on Hand stock if unforecasted demand is delivered while waiting for arrival.")
     daily_demand = fields.Float(
         "Daily Demand", digits='Product Unit', required=True, default=0.0, readonly=False, store=True)
-    based_on = fields.Selection(
+    min_max_based_on = fields.Selection(
         selection=[
             ('one_week', "Last 7 days"),
             ('one_month', "Last 30 days"),
@@ -84,7 +84,7 @@ class StockWarehouseOrderpoint(models.Model):
         help="Estimate the daily average future demand volume based on past period or choose Custom Demand to enter manually average daily demand.",
         required=True
     )
-    percent_factor = fields.Integer(default=100, required=True)
+    min_max_based_on_factor = fields.Integer(default=100, required=True)
     allowed_replenishment_uom_ids = fields.Many2many('uom.uom', compute='_compute_allowed_replenishment_uom_ids')
     replenishment_uom_id = fields.Many2one(
         'uom.uom', 'Multiple',
@@ -164,7 +164,7 @@ class StockWarehouseOrderpoint(models.Model):
         # We have to filter by company here in case of multi-company and because horizon_days is a company setting
         for company in orderpoints_to_compute.company_id:
             company_orderpoints = orderpoints_to_compute.filtered(lambda c: c.company_id == company)
-            horizon_date = fields.Date.today() + relativedelta.relativedelta(days=company_orderpoints.get_horizon_days())
+            horizon_date = fields.Date.today() + relativedelta(days=company_orderpoints.get_horizon_days())
             _, domain_move_in, domain_move_out, __, __ = company_orderpoints.product_id._get_domain_locations()
             domain_move_in = Domain.AND([
                 [('product_id', 'in', company_orderpoints.product_id.ids)],
@@ -199,7 +199,7 @@ class StockWarehouseOrderpoint(models.Model):
                 for move_date, move_qty in sorted(moves_by_product_dict.get((orderpoint.product_id.id, orderpoint.location_id.id), {}).items()):
                     qty_on_hand_at_date += move_qty
                     if qty_on_hand_at_date < orderpoint.product_min_qty:
-                        tentative_deadline = move_date - relativedelta.relativedelta(days=orderpoint.lead_days)
+                        tentative_deadline = move_date - relativedelta(days=orderpoint.lead_days)
                         break
                 orderpoint.deadline_date = tentative_deadline if tentative_deadline < horizon_date else False
 
@@ -209,7 +209,7 @@ class StockWarehouseOrderpoint(models.Model):
         for orderpoint in orderpoints_to_compute:
             values = orderpoint._get_lead_days_values()
             lead_days, _ = orderpoint.rule_ids._get_lead_days(orderpoint.product_id, bypass_delay_description=True, **values)
-            orderpoint.lead_horizon_date = fields.Date.today() + relativedelta.relativedelta(days=lead_days['total_delay'] + lead_days['horizon_time'])
+            orderpoint.lead_horizon_date = fields.Date.today() + relativedelta(days=lead_days['total_delay'] + lead_days['horizon_time'])
             orderpoint.lead_days = lead_days['total_delay']
         (self - orderpoints_to_compute).lead_horizon_date = False
         (self - orderpoints_to_compute).lead_days = 0
@@ -605,7 +605,7 @@ class StockWarehouseOrderpoint(models.Model):
             products = self.env['product.product'].browse(prod_ids)
             qties = products.with_context(
                 location=loc.id,
-                to_date=today + relativedelta.relativedelta(days=days)
+                to_date=today + relativedelta(days=days)
             ).read(['virtual_available'])
             for (product, qty) in zip(products, qties):
                 if product.uom_id.compare(qty['virtual_available'], 0) < 0:
@@ -856,3 +856,51 @@ class StockWarehouseOrderpoint(models.Model):
         - the value set on the company of the user if all else fail.
         """
         return self.env.context.get('global_horizon_days', (self.company_id or self.env.company).horizon_days)
+
+    def _get_period_of_time(self, period=None):
+        self.ensure_one()
+        if not period:
+            period = self.min_max_based_on
+        today = fields.Datetime.now()
+        start_date = limit_date = today
+        if period == 'one_week':
+            start_date = start_date - relativedelta(weeks=1)
+        elif period == 'one_month':
+            start_date = start_date - relativedelta(days=30)
+        elif period == 'three_months':
+            start_date = start_date - relativedelta(months=3)
+        elif period == 'one_year':
+            start_date = start_date - relativedelta(years=1)
+        else:  # Relative period of time.
+            start_date = datetime(year=today.year - 1, month=today.month, day=1)
+            if period == 'last_year_2':
+                start_date += relativedelta(months=1)
+            elif period == 'last_year_3':
+                start_date += relativedelta(months=2)
+            if period == 'last_year_quarter':
+                limit_date = start_date + relativedelta(months=3)
+            else:
+                limit_date = start_date + relativedelta(months=1)
+        return start_date, limit_date
+
+    def _get_daily_demand(self, period=None, ratio=None):
+        self.ensure_one()
+        if not ratio:
+            ratio = self.min_max_based_on_factor
+        date_from, date_to = self._get_period_of_time(period=period)
+        domain = Domain.AND([
+            [('product_id', '=', self.product_id.id)],
+            [('date', '>=', date_from)],
+            [('date', '<=', datetime.combine(date_to, time.max))],
+            [('state', 'in', ['assigned', 'confirmed', 'partially_available', 'done'])],
+            [('company_id', '=', self.company_id.id)],
+        ])
+        quantity_out = self.env['stock.move']._read_group(
+            Domain.AND([domain, [('location_dest_id.usage', 'in', ['customer', 'production'])]]),
+            aggregates=['product_qty:sum'],
+        )[0][0] or 0.0
+        quantity_returned = self.env['stock.move']._read_group(
+            Domain.AND([domain, [('location_id.usage', '=', 'customer')]]),
+            aggregates=['product_qty:sum'],
+        )[0][0] or 0.0
+        return ((quantity_out - quantity_returned) / (date_to - date_from).days) * (ratio / 100)
